oidc_connect.go

```diff
@@ -0,0 +1,94 @@
+package keycloak
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	errorhandler "github.com/cloudtrust/common-service/errors"
+	"github.com/cloudtrust/common-service/log"
+)
+
+// OidcTokenProvider provides OIDC tokens
+type OidcTokenProvider interface {
+	ProvideToken(ctx context.Context) (string, error)
+}
+
+type oidcToken struct {
+	AccessToken      string `json:"access_token,omitempty"`
+	ExpiresIn        int64  `json:"expires_in,omitempty"`
+	RefreshToken     string `json:"refresh_token,omitempty"`
+	RefreshExpiresIn int64  `json:"refresh_expires_in,omitempty"`
+	TokenType        string `json:"token_type,omitempty"`
+	NotBeforePolicy  int    `json:"not-before-policy,omitempty"`
+	SessionState     string `json:"session_state,omitempty"`
+	Scope            string `json:"scope,omitempty"`
+}
+
+type oidcTokenProvider struct {
+	timeout    time.Duration
+	tokenURL   string
+	reqBody    string
+	logger     log.Logger
+	oidcToken  oidcToken
+	validUntil int64
+}
+
+// NewOidcTokenProvider creates an OidcTokenProvider
+func NewOidcTokenProvider(config Config, realm, username, password, clientID string, logger log.Logger) OidcTokenProvider {
+	var tokenURL = fmt.Sprintf("%s/auth/realms/%s/protocol/openid-connect/token", config.AddrAPI, realm)
+	// If needed, can add &client_secret={secret}
+	var body = fmt.Sprintf("grant_type=password&client_id=%s&username=%s&password=%s",
+		url.QueryEscape(clientID), url.QueryEscape(username), url.QueryEscape(password))
+
+	return &oidcTokenProvider{
+		timeout:  config.Timeout,
+		tokenURL: tokenURL,
+		reqBody:  body,
+		logger:   logger,
+	}
+}
+
+func (o *oidcTokenProvider) ProvideToken(ctx context.Context) (string, error) {
+	if o.validUntil > time.Now().Unix() {
```

Comment:
Maybe we should have a margin of a few seconds to avoid expiration between this evaluation and call to KC ?

Reply:
I hard coded a 5 seconds delay... not sure it is really important to configure this delay